src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostOnAutoInsertEndpointTest.cs

```diff
@@ -0,0 +1,136 @@
+    }
+
+    [Theory]
+    [InlineData("""
+            @code {
+                //$$
+                void TestMethod() {}
+            }
+            """,
+        "/// <summary>", "/")]
+    public async Task Component_AutoInsertCSharp(string input, string insertedText, string triggerCharacter)
+    {
+        await VerifyOnAutoInsertAsync(input, insertedText, triggerCharacter);
+    }
+
+    private async Task VerifyOnAutoInsertAsync(
+        string input,
+        string insertedText,
+        string triggerCharacter,
+        bool createDelegatedResponse = false)
+    {
+        TestFileMarkupParser.GetPosition(input, out input, out var cursorPosition);
+        var document = CreateProjectAndRazorDocument(input);
+        var sourceText = await document.GetTextAsync(DisposalToken);
+
+        var clientSettingsManager = new ClientSettingsManager([], null, null);
+        clientSettingsManager.Update(ClientAdvancedSettings.Default with { FormatOnType = true, AutoClosingTags = true });
+
+        IOnAutoInsertTriggerCharacterProvider[] onAutoInsertTriggerCharacterProviders = [
+            new RemoteAutoClosingTagOnAutoInsertProvider(),
+            new RemoteCloseTextTagOnAutoInsertProvider()];
+
+        VSInternalDocumentOnAutoInsertResponseItem? response = null;
+        if (createDelegatedResponse)
+        {
+            var start = sourceText.GetPosition(cursorPosition + triggerCharacter.Length);
+            var end = start;
+            response = new VSInternalDocumentOnAutoInsertResponseItem()
+            {
+                TextEdit = new TextEdit() { NewText = insertedText, Range = new() { Start = start, End = end } },
+                TextEditFormat = InsertTextFormat.Snippet
+            };
+        }
+
+        var requestInvoker = new TestLSPRequestInvoker([(VSInternalMethods.OnAutoInsertName, response)]);
+
+        var endpoint = new CohostOnAutoInsertEndpoint(
+            RemoteServiceInvoker,
+            clientSettingsManager,
+            onAutoInsertTriggerCharacterProviders,
+            TestHtmlDocumentSynchronizer.Instance,
+            requestInvoker,
+            LoggerFactory);
+
+        var formattingOptions = new FormattingOptions()
+        {
+            InsertSpaces = true,
+            TabSize = 4
+        };
+
+        var request = new VSInternalDocumentOnAutoInsertParams()
+        {
+            TextDocument = new TextDocumentIdentifier()
+            {
+                Uri = document.CreateUri()
+            },
+            Position = sourceText.GetPosition(cursorPosition),
+            Character = triggerCharacter,
+            Options = formattingOptions
+        };
+
+        var result = await endpoint.GetTestAccessor().HandleRequestAsync(request, document, DisposalToken);
+
+        Assert.NotNull(result);
+
+        if (createDelegatedResponse)
+        {
+            Assert.Equal(response, result);
+        }
+
+        Assert.Equal(insertedText, result.TextEdit.NewText);
```

Comment:
Rather than validating responses directly, it would be better to apply the result edit to the original document, and have the tests provide the expected end state of the document. That way if something changes, like we start returning smaller edits or multiple edits or something, the tests don't break. Seeing the document final state also makes the tests much more readable.